### ISSUE TITLE
LFX Workspace: Integrate MLX as a new WASI-NN backend

### ISSUE THREAD

**Issue by guptaaryan16** [WasmEdge/WasmEdge#3266]
## Motivation:

[**MLX**](http://github.com/ml-explore/mlx) has been a significant development in the field of ML, especially with the popularity of M1 series Apple Macs, as they propose a unified memory architecture for CPUs and GPUs on an SoC-based chip, which not only improves speed but also the training time and other costs. As **WasmEdge** becomes famous as a Wasm runtime, it seems only essential to support the MLX framework directly on Wasm backends and take advantage of memory architecture on the M1 macs.

## Milestones for the project

[ Week 1-3 ]
- Complete review of C++ literature and debugging techniques (from Federico Busato/ Modern C++ lectures) along with review and experiments on the WasmEdge and MLX codebases for integration
- Complete the plugin files `mlx.cpp` and `mlx.h` files using references from TensorFlow, PyTorch and GGML plugins.
- Complete the first (draft) PR with some basic functionality and provide some model inference capabilities for a basic LLM model using `mlx-llm` project

Some initial files for the plugin are [here](https://github.com/guptaaryan16/LFX-WasmEdge-MLX-Pre-Task/tree/main/plugin_sample)

[ Week 3-6 ]
- Debug the initial files and provide benchmarks for the LLM models run w.r.t GGML plugin
-  Complete the unit tests on the initial plugin
- Complete the files and get a review in a draft PR 
- Complete the merging of the first PR (with some CI tests) (try to complete it a week before the deadline, i.e., 10th April)

[ Week 6-11 ]
- Complete unit tests and test all the functionality for the MLX based inference on Apple GPUs
- Able to compile and use examples from the `mlx-c` project to web assembly using WasmEdge runtime and directly use them for the LLM and other small-scale web inference
- Complete the docs for the plugin and provide some tutorials for its usage (mainly in C++ for now; let's hope to complete support more languages like C, Rust etc. in the future if possible)
- (optional) Provide some functionality to be able to finetune models based on 'mlx-llm' frameworks and plugins and also allow MLX objects(like arrays, nn, functions) to be used and compiled in Wasm,  write tests for all the main features of MLX, including arrays, nn, transformers, grad, map, etc.

Last week as buffer for unexpected delay.

**Note:** 
I may be a little busy during my exams from 24th April to 3rd May 2024, so I will adjust the development hours before and after these dates to complete all necessary milestones.

Also, I would like to discuss some of these features with the MLX community to get much-needed ideas and feedback from their side.
cc @awni

**Comment by guptaaryan16** [comment]
### Week 1:

- Reading and brushing up on C++ basics and standards from Modern C++ programming(Federico Busato)
- Reading the [plugin docs](https://wasmedge.org/docs/contribute/plugin/develop_plugin_cpp) on the WasmEdge website 
- Initial Ideas from the `ml-explore/mlx-c project` from the Apple ecosystem
- Completing the initial workspace issue in the WasmEdge repo

### Week 2
- Completed all lectures and some programs in C++ and cmake to understand the behaviour of things within the ecosystem, Also use Valgrind and leaks to understand some debugging 
- Discussed with `hydai` and `awni` about usage of either safetensors or ggml load in mlx/io
- Wrote small programs and researched to understand C++ behaviour 
- Will start with the plugin this week(probably soon) after understanding a bit about what designs of previous plugins are

**Comment by guptaaryan16** [comment]
## Week 3-4
- Since on inspection and discussion with mentors(hydai and awni), I did not find any implementation for C++ based inference on MLX/C++ directly for LLAMA and other llms, so I started working on my own NN class and definitions and will try to complete a working example by sunday(end of 4th week)
- I have also worked on the plugin files and completed the initial loading of model weights part (using API from MLX and WasiNN plugin) -> `load` function
- Working on the CPP frontend for the MLX library to load weights and use them in the layers API
- Research and complete the working example of CPP based on the MLX C++ API

## Week 5-6 
- Completed Initial implementation of NN Classes and working on loading weights problem(still WIP) with the custom NN classes

**Comment by hydai** [comment]
@guptaaryan16 
Please open a working PR related to your work ASAP for the midterm evaluation.

**Comment by hydai** [comment]
Please ensure that you post the progress every week.

**Comment by guptaaryan16** [comment]
Hi @hydai sorry for not updating my progress over here:

## Week 7-8
- Currently working on implementing the classes and functions related to LLAMA models with thorough testing of inputs and outpus 
- Was able to load weights from a GGUF type and safetensor type files and tensors correctly(finally)
- Need to still work on `register_layer` function for adding layers to the model and thus create a workable implementation of TransformerBlock
- Aiming to complete a workable implementation by the end of the week
- Also trying to understand some parts of `LLAMA.cpp` for usage in our project if possible

Also I would like to communicate that for the end of this week and next week, I may work for less time due to my university exams but would cover the work after the exams are over (in the following weeks).

**Comment by guptaaryan16** [comment]
Week 9-  10
- Completed work on `register_layer`, `load_from_GGUF` and `load_from_safetensors` functions
- Completing two model files `phi2.cpp` and `llama.cpp` using the classes and implemented necessary parts like `Embedding`, `PhiAttention` etc. as necessary, still working on some bugs related to embeddings and tokenisation
- After this, I am gonna work on the mlx plugin files and complete them before the deadline, adding all these functionalities and creating a working demo  
- Also will start work on documentation from today so that it gets completed within deadline

**Comment by hydai** [comment]
It's very close to the deadline, please push your commits to the PR.

**Comment by hydai** [comment]
This is a final notification. We still haven't received your PR or updates for the final evaluation.

**Comment by guptaaryan16** [comment]
Hi @hydai I am completing some work on the PR and hope to submit it as soon as possible, sorry for the delay.

**Comment by hydai** [comment]
Please push working commits, even if they're dirty, to update the current PR so we can at least base the final evaluation on that.

**Comment by hydai** [comment]
Well received. Thanks @guptaaryan16 
Happy coding :-)
